library/src/blas1/rocblas_scal.hpp

```diff
@@ -0,0 +1,107 @@
+/* ************************************************************************
+ * Copyright 2016-2019 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+#include "handle.h"
+#include "rocblas.h"
+#include "utility.h"
+
+template <typename T, typename U, typename V>
+__global__ void rocblas_scal_kernel(rocblas_int    n,
+                                    V              alpha_device_host,
+                                    rocblas_stride stride_alpha,
+                                    U              xa,
+                                    rocblas_int    offsetx,
+                                    rocblas_int    incx,
+                                    rocblas_stride stridex)
+{
+    T*        x     = load_ptr_batch(xa, hipBlockIdx_y, offsetx, stridex);
+    auto      alpha = load_scalar(alpha_device_host, hipBlockIdx_y, stride_alpha);
+    ptrdiff_t tid   = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+
+    // bound
+    if(tid < n)
+        x[tid * incx] *= alpha;
+}
+
+template <rocblas_int NB, typename T, typename U, typename V>
+rocblas_status rocblas_scal_template(rocblas_handle handle,
+                                     rocblas_int    n,
+                                     const V*       alpha,
+                                     rocblas_stride stride_alpha,
+                                     U              x,
+                                     rocblas_int    offsetx,
+                                     rocblas_int    incx,
+                                     rocblas_stride stridex,
+                                     rocblas_int    batch_count,
+                                     V*             mem)
+{
+    // Memory queries must be in template as _impl doesn't have stride_alpha parameter (for calls from
+    // outside of rocblas)
```

Comment:
> // Memory queries must be in template as _impl doesn't have stride_alpha parameter (for calls from
>  // outside of rocblas)

However, this prevents `rocblas_scal_template` from being called from any other _arbitrary_ rocBLAS routine.

Only one device memory allocation is allowed to be active per rocBLAS C API call, and it must allocate for itself and all of its children.

If another routine were to allocate device memory through rocBLAS, directly or indirectly, then it could not call `rocblas_scal_template`, unless that's the only routine in its call tree which allocates device memory, or unless it frees its allocated memory before calling `rocblas_scal_template`. 

 (`hipMalloc()` and `hipFree()` should not be called from the rocBLAS library except in certain ways, because they are synchronizing operations.)

There are runtime safeguards in the library to detect violations of these restrictions, with an internal error message and an `abort()`.

As long as the test pass right now, the error is not happening, but beware of this restriction.